ssl: rudimentary decryption for TLS 1.2

1wilkens opened the pull request:
Hi,

this is our draft implementation for transparent TLS 1.2 decryption in Zeek. It supports both TLS pre-master secrets and TLS session keys (and IVs). Although several limitations still exist (see below), we wanted to post it here for review.

Thank you @0xxon for your comments and help up until now :)

Several limitations still apply:
- TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 is the only supported cipher
suite
- No zeek tests (the patch was tested extensively for our paper)
- Several small open technical questions marked with FIXME
- Architecture in the ssl module might not be optimal

1wilkens commented:
Seems like the failing platforms do not have `<openssl/kdf.h>`. Should key derivation then be feature-gated via a preprocessor macro? Also the sanitizer reports some data leakage that is also related to our usage of client and server random in `Handshake_Conn`.

0xxon commented:
Hmmmm...

we typically try to avoid having such gated features -- however, for this I could potentially see it.

Is it possible that this functionality is just in another header file in older OpenSSL releases - or do they just plainly not have the feature at all?

Apart from that - would it be possible for you to add a test-case for this new code? We do require that fir basically all new features, and I think that it should be totally be possible to add a testcase for this based on the traffic that you did generate for the paper :)

I also apologize in advance - it might take me a few days to look over the source-code. And given that this interacts with a few relatively core components (the SSL analyzer is widely tested), this might take a bit of iteration till it can make it in. This is, however, a really neat feature - thank you for pushing this :)

0xxon commented on scripts/policy/protocols/ssl/decryption.zeek:
```diff
@@ -0,0 +1,106 @@
+		client_random: string &log &optional;
+	};
+
+	const keylog_file = getenv("ZEEK_TLS_KEYLOG_FILE") &redef;
```
just to check - did you ever end up implementing the keys back to Zeek via broker? Or did you only use the keylog file?

Edit: I see found the broker bit later. This probably should be docmented here.

The current approach of key ingestion also will not work in a cluster setting - is that something that you might be interesting of fixing?

1wilkens replied:
You are correct, currently the keys can either be received by broker or by ingesting a slightly modified keyfile on launch. Is there an equivalent of the Input Framework for cluster mode? Also what would it take to make key storage cluster compatible? Can we just store the secrets in the proxy?

0xxon commented on scripts/policy/protocols/ssl/decryption.zeek:
```diff
@@ -0,0 +1,106 @@
+module SSL;
+
+# Local
+const input_stream_name = "input-tls-keylog-file";
```
Is there a reason to have this as a const?

1wilkens replied:
Not really, I just didn't see a use-case for making this &redef.

0xxon commented on scripts/policy/protocols/ssl/decryption.zeek:
```diff
@@ -0,0 +1,106 @@
+
+global randoms: table[string] of string = {};
+
+export {
```
Small style not - we typically have the export section before any other type declarations. Though I can just fix that while merging

1wilkens replied:
I tried to do that across the codebase, but I seem to have missed some places. I will make another pass soon and resolve this.

0xxon commented on scripts/policy/protocols/ssl/decryption.zeek:
```diff
@@ -0,0 +1,106 @@
+	global keys: table[string] of string = {} &redef;
+
+	event SSL::add_keys(client_random: string, val: string)
+	{
```
The braces here should indented one more level (with the following source-code). Also - style-nit: we typically don't put full events in the export sections - we  just declare them here and then add the event later.

1wilkens replied:
I tried to do the formatting across the codebase, but I missed some places. I will do another pass over the whole patch and make that consistent. Events I'll just change as per your suggestion.

0xxon commented on scripts/policy/protocols/ssl/decryption.zeek:
```diff
@@ -0,0 +1,106 @@
+event Input::end_of_data(name: string, source: string)
+	{
+	if ( name == input_stream_name )
+	{
```
braces should be indented one more level again.

0xxon commented on scripts/policy/protocols/ssl/decryption.zeek:
```diff
@@ -0,0 +1,106 @@
+		else
+			{
+			# FIXME: should this be moved to reporter.log or removed completely?
+			#print "No suitable key or secret found for random:", randoms[c$uid];
```
probably not worth reporting. If at all - one could choose to have a reporter output here that is gated behind an @if, so one can elect to activate it as debug output without it having runtime performance impact.

0xxon commented on scripts/policy/protocols/ssl/decryption.zeek:
```diff
@@ -0,0 +1,106 @@
+	# ingest keylog file if the environment is set
+	if ( keylog_file != "" )
+		{
+		suspend_processing();
```
This probably is here due to the test-setup that you used - but we should definitely not ship policy scripts that have a chance of calling suspend_processing :)

1wilkens replied:
Yeah that was the only way I could let zeek ingest the keylog file completely before processing a provided pcap. If that is not a common use-case, we can remove the Input Framework part completely I guess.

0xxon commented on scripts/policy/protocols/ssl/decryption.zeek:
```diff
@@ -0,0 +1,106 @@
+		{
+		suspend_processing();
+
+		Input::add_table([$name=input_stream_name, $source=keylog_file, $destination=secrets, $idx=Idx, $val=Val, $want_record=F]);
```
I am not sure if you did think about this - but it also is possible to just open the file in streaming mode and to keep it open - in that case, new entries will be added to the table automatically when they arrive.

1wilkens replied:
I did not know that. Our use-case just required to read all keys upfront, but that could be interesting. However, the keylog needs to be preprocessed to be read by zeek anyway so I am not sure, if we need this feature at all.

0xxon commented on src/analyzer/protocol/ssl/DTLS.cc:
```diff
@@ -72,4 +72,9 @@ void DTLS_Analyzer::SendHandshake(uint16_t raw_tls_version, uint8_t msg_type, ui
 
+bool DTLS_Analyzer::TryDecryptApplicationData(int len, const u_char* data, bool is_orig, uint8_t content_type, uint16_t raw_tls_version)
+	{
+		// noop for now as DTLS decryption is currently not supported
```
indentation of this is off again :)

0xxon commented on src/analyzer/protocol/ssl/SSL.cc:
```diff
@@ -16,12 +84,20 @@ SSL_Analyzer::SSL_Analyzer(Connection* c)
 	had_gap = false;
+	c_seq = 0;
+	s_seq = 0;
+	// FIXME: should this be initialized to nullptr?
```
yup :)

0xxon replied:
Also - for extra points - we try to use intrusive-pointers in new code. 

https://github.com/zeek/zeek/wiki/IntrusivePtr-Overview

0xxon commented on src/analyzer/protocol/ssl/SSL.cc:
```diff
@@ -8,6 +9,73 @@
+#define MSB(a) ((a>>8)&0xff)
+#define LSB(a) (a&0xff)
+
+static void fmt_seq(uint32_t num, u_char* buf)
```
these probably all should end up in the zeek::analyzer::ssl namespace, unless there is a big reason to not have them there.

0xxon commented on src/analyzer/protocol/ssl/SSL.cc:
```diff
@@ -8,6 +9,73 @@
+	printf("\n");
+	}
+
+static bool tls1_prf(const u_char *secret, int secret_len, const char* label,
```
This probably should be a member function of the analyzer - unless there is a specific reason why it shouldn't be.

0xxon commented on src/analyzer/protocol/ssl/SSL.cc:
```diff
@@ -8,6 +9,73 @@
+	// alloc buffers
+	EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_TLS1_PRF, NULL);
+	size_t seed_len = strlen(label) + rnd1_len + rnd2_len;
+	u_char *seed = (u_char*) malloc(seed_len);
```
could we use c++-style allocations here? :)

0xxon commented on src/analyzer/protocol/ssl/SSL.cc:
```diff
@@ -8,6 +9,73 @@
+	u_char *ptr = seed;
+
+	// seed = label + rnd1 + rnd2
+	memcpy(ptr, label, strlen(label));
```
Just to check - is it guaranteed that the label does not contain 0-bytes? Also - can this potentially just be done with std::strings - all of this is mostly string-concatenations?

0xxon commented on src/analyzer/protocol/ssl/SSL.cc:
```diff
@@ -98,4 +174,166 @@ void SSL_Analyzer::Undelivered(uint64_t seq, int len, bool orig)
+	{
+	// FIXME: Is this the proper way to initialize a zeek::StringVal?
+	//        the alternative requires a cast: new zeek::StringVal(len, data)
+	secret = new zeek::StringVal(new zeek::String(data, len, true));
```
Optimally, it would be neat to make this an IntrusivePtr - then you don't have to worry about cleaning it up - or this function being called several times.

Apart from that - personally I would prefer the cast from u_char to char, without goint through zeek::String.

0xxon commented on src/analyzer/protocol/ssl/SSL.cc:
```diff
@@ -98,4 +174,166 @@ void SSL_Analyzer::Undelivered(uint64_t seq, int len, bool orig)
+	// Neither secret or key present: abort
+	if ( secret->Len() == 0 && keys->Len() == 0 )
+		{
+		// FIXME: this is just for debugging
```
those printfs should obviously go. It would be neat to instead make them into debug messages, so e.g....

```
DBG_LOG(DBG_ANALYZER, "Could not decrypt packet...")
```

0xxon commented on src/analyzer/protocol/ssl/SSL.cc:
```diff
@@ -98,4 +174,166 @@ void SSL_Analyzer::Undelivered(uint64_t seq, int len, bool orig)
+	}
+
+	// Keys present: decrypt TLS application data
+	if ( keys->Len() != 0 )
```
this probably also should check that keys has the right length? I know that that is given at the moment - but I am a fan of defensive programming :)

0xxon commented on src/analyzer/protocol/ssl/SSL.cc:
```diff
@@ -98,4 +174,166 @@ void SSL_Analyzer::Undelivered(uint64_t seq, int len, bool orig)
+	if ( keys->Len() != 0 )
+		{
+		// session keys & AEAD data
+		u_char c_wk[32];
```
it would be neat to label the non-obvious ones of these with a comment.

1wilkens replied:
Done

0xxon commented on src/analyzer/protocol/ssl/SSL.cc:
```diff
@@ -98,4 +174,166 @@ void SSL_Analyzer::Undelivered(uint64_t seq, int len, bool orig)
+		u_char* encrypted = (u_char*)data;
+		size_t encrypted_len = len;
+
+		// FIXME: should this be moved to SSL_Conn?
```
I actually am not sure - but am tempted to leave this here for now.

0xxon commented on src/analyzer/protocol/ssl/SSL.cc:
```diff
@@ -98,4 +174,166 @@ void SSL_Analyzer::Undelivered(uint64_t seq, int len, bool orig)
+		int res = 0;
+		if ( ! (res = EVP_DecryptFinal(ctx, NULL, &res)) )
+			{
+			printf("Decryption failed with return code %d. Invalid key?\n", res);
```
This one again seems great to be a debug log statement.

1wilkens replied:
Agreed! Is there a dedicated Zeek function for that?

0xxon replied:
There actually is, yes. Search the codebase for DBG_LOG :)

1wilkens replied:
Ah I didn't see you comments above ;)

0xxon commented on src/analyzer/protocol/ssl/SSL.h:
```diff
@@ -33,11 +34,24 @@ class SSL_Analyzer final : public analyzer::tcp::TCP_ApplicationAnalyzer {
 	static analyzer::Analyzer* Instantiate(Connection* conn)
 		{ return new SSL_Analyzer(conn); }
 
+	// Key material for decryption
```
Since these are new functions, we should follow our general, more verbose documentation guidelines. See e.g. how the functions in https://github.com/zeek/zeek/blob/master/src/analyzer/Manager.h are documented.

1wilkens replied:
Yeah, documentation is definitely not there yet. I'll come up with something on the next pass.

1wilkens replied:
Addressed in the latest version

0xxon commented on src/analyzer/protocol/ssl/CMakeLists.txt:
```diff
@@ -10,8 +10,8 @@ zeek_plugin_bif(events.bif)
 zeek_plugin_pac(tls-handshake.pac tls-handshake-protocol.pac tls-handshake-analyzer.pac ssl-defs.pac
-	proc-client-hello.pac
-	proc-server-hello.pac
+	proc-client-hello-tls.pac
```
I assume I am lazy here - is it absolutely necessary to split these up?

1wilkens replied:
The only diff in the `-tls` versions is, that we store `client_random` and `server_random` in the connection which only applies for TLS handshake connections but not SSL connections. I assumed that the files only existed to not duplicate the identical methods between SSL and TLS. However, as they are now not identical anymore, we could also just move the definitions into the respective `*-analyzer.pac` files and remove the extra files.

0xxon replied:
Yup, that would probably make more sense. Especially since we then also can throw out some of the logic that checks if this was an SSL or a TLS event.

1wilkens replied:
Done, although I did not find any logic related to SSL/TLS checks directly in the files. They might be somewhere else though?

1wilkens commented:
I'll reduce on of our captured pcaps to have a small testcase. Regarding openSSL: I have not found an API to do TLS PRF in versions <1.1.0 but there should be something. Since the decryption API is already there in lower versions, we could also feature gate just the key derivation and only support decryption with derived session keys in that case.

0xxon commented:
> I'll reduce on of our captured pcaps to have a small testcase. 

Thank you

> Regarding openSSL: I have not found an API to do TLS PRF in versions <1.1.0 but there should be something. Since the decryption API is already there in lower versions, we could also feature gate just the key derivation and only support decryption with derived session keys in that case.

Ok, fair

> EDIT: Also another style question: Do you prefer a single squashed commit per PR or incremental commits for missing stuff like adding a testcase and so on?

In cases of significant additions, at least I actually prefer incremental commits - especially if they help with understanding how things were developed/added. So - I would not bother with squashing at the moment - if it gets much too large you still can do that at the end.

1wilkens commented:
Addresses some of the comments, mostly style and structure related ones. Still open are:
- tests(!)
- documentation
- feature-gating key derivation
- key reading via input framework
- cluster mode
- some more defensive checks (key length etc.) and minor logging things

I converted the TLS PRF to use `std::string` and overall tried to be a bit more C++ ideomatic. As my C++ is not great, another review especially for the PRF would be great @0xxon. I'll address the issues above in the next passes, but let me know if I have missed something.

1wilkens commented:
I updated the branch with the changes from zeek/cmake#29 and now it compiles on all operating systems. However, as key derivation is not present on the ones without the openssl header, the test fails. Is there a way to conditionally disable tests? Otherwise I'd have to change the test to use derived keys instead of pre-master secrets as pure decryption should work in all configurations.

0xxon commented:
> I updated the branch with the changes from [zeek/cmake#29](https://github.com/zeek/cmake/pull/29) and now it compiles on all operating systems. However, as key derivation is not present on the ones without the openssl header, the test fails. Is there a way to conditionally disable tests? Otherwise I'd have to change the test to use derived keys instead of pre-master secrets as pure decryption should work in all configurations.
> 

There is a way to disable tests, yes. You can use `@TEST-REQUIRES` (see https://github.com/zeek/btest) to skip a test is a requirement is not met. There are a few examples that use that in our test suite. Does that work for what you need?

Is this prretty much done apart from this - and besides the memory-leaks? I probably won't have much time to look into this this week - but I should have a bit of time next week to review and see how we continue :)

> EDIT: And there seems to be a memory leak when setting the client and server random. That is probably related to my missing knowledge about how to initialize `binpac::datastring`s, so it probably needs some more work.

1wilkens commented:
Sorry for the late reply!

I updated the PR to use the merged branch from `cmake` and feature-gated the tests based on `OPENSSL_HAVE_KDF_H`. I also added documentation for the new methods in the analyzer and updated the data structures for secret and key storage to something more suitable instead of `zeek::StringVal`. You might want to review these changes, as C++ is still not my strong suite ;)

Beyond that the only real open issue would be dealing with the cluster setup. I have not changed anything in this regard, as I am not sure how to proceed there as mentioned previously.

If there are any other changes I can implement, let me know!
Also great to see this planned for 4.2 for some additional motivation. :)

1wilkens commented:
Seems like I managed to chase down the missing `free`s and the sanitizer is happy now. However, debian9 now randomly fails but that looks like an unrelated one-off failure.

timwoj commented:
Looks like a transient failure to me too. I restarted the job.

1wilkens commented:
So CI is green and the only open questions is dealing with the cluster framework and the usage of the input framework for reading secrets on start.
@0xxon any ideas how to proceed? If the changes to adapt for cluster mode are mostly straightforward, I can try to come up with something.

0xxon commented:
Let me give this another pass next week. Sorry for the delay - the last weeks were a bit busy.

0xxon commented:
Also - is it ok to just push small changes to this branch, or should I put them in another branch first?

1wilkens commented:
Sure just push to this branch
